File: platforms/ios/Dezel/DezelTests/Core/JavaScriptContextTest.swift
import XCTest
import Dezel

class JavaScriptContextTest: XCTestCase {

	internal static var testJavaScriptFunctionCallback: JavaScriptFunctionHandler? = nil
	internal static var testConstructorACallback: JavaScriptFunctionHandler? = nil
	internal static var testConstructorBCallback: JavaScriptFunctionHandler? = nil
	internal static var testJavaScriptGetterCallback: JavaScriptGetterHandler? = nil
	internal static var testJavaScriptSetterCallback: JavaScriptSetterHandler? = nil

	private var context: JavaScriptContext = JavaScriptContext()

	override func setUp() {
		super.setUp()
		self.context.dispose()
		self.context = JavaScriptContext()
		self.context.setup()
	}

	func testValueCreateWithNull() {
		let value = self.context.createNull()
		XCTAssertTrue(value.isNull)
		value.dispose()
	}

	func testValueCreatedWithUndefined() {
		let value = self.context.createUndefined()
		XCTAssertTrue(value.isUndefined)
		value.dispose()
	}

	func testValueCreatedWithBoolean() {
		let value = self.context.createBoolean(true)
		XCTAssertTrue(value.isBoolean)
		XCTAssertEqual(value.boolean, true)
		value.dispose()
	}

	func testValueCreatedWithString() {
		let value = self.context.createString("foo")
		XCTAssertTrue(value.isString)
		XCTAssertEqual(value.string, "foo")
		value.dispose()
	}

	func testValueCreatedWithNumber() {
		let value = self.context.createNumber(10.0)
		XCTAssertTrue(value.isNumber)
		XCTAssertEqual(value.number, 10.0)
		value.dispose()
	}

	func testValueCreatedWithObject() {
		let value = self.context.createEmptyObject()
		XCTAssertTrue(value.isObject)
		value.dispose()
	}

	func testValueCreatedWithFunction() {
		let value = self.context.createFunction({ (callback: JavaScriptFunctionCallback) in })
		XCTAssertTrue(value.isFunction)
		value.dispose()
	}

	func testEvaluation() {
		self.context.evaluate("1 + 2")
	}

	func testGlobalVariables() {

		let object = self.context.createEmptyObject()
		self.context.global.property("value", value: object)
		let result = self.context.global.property("value")
		XCTAssertTrue(object.equals(result))
		result.dispose();
		object.dispose()

		self.context.global.property("value", string: "test")
		XCTAssertEqual(self.context.global.property("value").string, "test")

		self.context.global.property("value", number: 10)
		XCTAssertEqual(self.context.global.property("value").number, 10)

		self.context.global.property("value", boolean: true)
		XCTAssertEqual(self.context.global.property("value").boolean, true)

		self.context.global.property("value", value: nil)
		XCTAssertEqual(self.context.global.property("value").string, "null")
	}

	func testJavaScriptClassCreation() {
		let classA = self.context.createClass(JavaScriptContextTestClassA.self)
		classA.dispose()
	}


	func testExceptionHandler() {

		var didCallExceptionHandler = false

		self.context.handleError { error in
			didCallExceptionHandler = true
		}

		self.context.evaluate("a + b")

		XCTAssertTrue(didCallExceptionHandler)
	}
}

File: platforms/android/Dezel/dezel/src/androidTest/java/ca/logaritm/dezel/core/JavaScriptContextTest.kt
package ca.logaritm.dezel.core

import ca.logaritm.dezel.ContextTestClassA
import ca.logaritm.dezel.ContextTestObject
import org.junit.Assert
import org.junit.Before
import org.junit.Test

class JavaScriptContextTest {

	private var context: JavaScriptContext = JavaScriptContext()
	private var expectedCallee: JavaScriptValue? = null
	private var expectedObject: JavaScriptValue? = null
	private var expectedContext: JavaScriptContext? = null

	@Before
	fun beforeTest() {
		this.context.dispose()
		this.context = JavaScriptContext()
		this.expectedCallee = null
		this.expectedObject = null
		this.expectedContext = null
	}

	@Test
	fun testValueCreateWithNull() {
		val value = this.context.createNull()
		Assert.assertTrue(value.isNull)
		value.dispose()
	}

	@Test
	fun testValueCreatedWithUndefined() {
		val value = this.context.createUndefined()
		Assert.assertTrue(value.isUndefined)
		value.dispose()
	}

	@Test
	fun testValueCreatedWithBoolean() {
		val value = this.context.createBoolean(true)
		Assert.assertTrue(value.isBoolean)
		Assert.assertEquals(value.boolean, true)
		value.dispose()
	}

	@Test
	fun testValueCreatedWithString() {
		val value = this.context.createString("foo")
		Assert.assertTrue(value.isString)
		Assert.assertEquals(value.string, "foo")
		value.dispose()
	}

	@Test
	fun testValueCreatedWithNumber() {
		val value = this.context.createNumber(10.0)
		Assert.assertTrue(value.isNumber)
		Assert.assertEquals(value.number, 10.0, 0.0)
		value.dispose()
	}

	@Test
	fun testValueCreatedWithObject() {
		val value = this.context.createEmptyObject()
		Assert.assertTrue(value.isObject)
		value.dispose()
	}

	@Test
	fun testValueCreatedWithFunction() {
		val value = this.context.createFunction(fun(callback: JavaScriptFunctionCallback) {})
		Assert.assertTrue(value.isFunction)
		value.dispose()
	}

	@Test
	fun testEvaluation() {
		this.context.evaluate("1 + 2")
	}

	@Test
	fun testGlobalVariables() {

		val obj = this.context.createEmptyObject()
		this.context.global.property("value", obj)
		val result = this.context.global.property("value")
		Assert.assertTrue(obj.equals(result))
		result.dispose()
		obj.dispose()

		this.context.global.property("value", "test_string")
		Assert.assertEquals(this.context.global.property("value").string, "test_string")

		this.context.global.property("value", 10.0)
		Assert.assertEquals(this.context.global.property("value").number, 10.0, 0.0)

		this.context.global.property("value", true)
		Assert.assertEquals(this.context.global.property("value").boolean, true)

		this.context.global.property("value", null)
		Assert.assertEquals(this.context.global.property("value").string, "null")
	}

	@Test
	fun testBridgedObjectCreation() {
		val objectA = this.context.createObject(ContextTestObject::class.java)
		objectA.dispose()
	}

	@Test
	fun testClassTemplateCreation() {
		val classA = this.context.createClass(ContextTestClassA::class.java)
		classA.dispose()
	}

	@Test
	fun testAttributes() {

		val value = Object()

		this.context.attribute("key1", value)

		Assert.assertEquals(this.context.attribute("key1"), value)
		Assert.assertEquals(this.context.attribute("key2"), null)

		this.context.attribute("key1", null)

		Assert.assertEquals(this.context.attribute("key1"), null)
	}

	@Test
	fun testExceptionHandler() {

		var didCallExceptionHandler = false

		this.context.handleError { error ->
			didCallExceptionHandler = true
		}

		this.context.evaluate("a + b")

		Assert.assertTrue(didCallExceptionHandler)
	}

	companion object {
		var testFunctionCallback: JavaScriptFunctionHandler? = null
		var testConstructorACallback: JavaScriptFunctionHandler? = null
		var testConstructorBCallback: JavaScriptFunctionHandler? = null
		var testPropertyGetterCallback: JavaScriptGetterHandler? = null
		var testPropertySetterCallback: JavaScriptSetterHandler? = null
	}
}
